src/box/lua/config/utils/schema.lua

```diff
@@ -125,6 +138,10 @@ end
 --     -- Return a parsed/typecasted value of the given scalar
 --     -- type otherwise.
 --     fromenv = <function>,
+--
+--     -- If any number value definitely can't pass validation
+--     -- against the given scalar, this property is true.
+--     never_accept_number = <boolean>,
```

Comment:
I don't like negative properties and these calculations in the head too. While working on the patch, I've tried to reverse these 'never accept number' and 'never accept a (non-empty) array' predicates, but things become worse.

The reason is that here we have a situation that is not just black/white, but the black/gray/white one. We're interested in a guaranteed property 'neither accepted input values has property X' and it is not a negation of 'accepted input values has property X' (that reads as 'all accepted', not 'some accepted').

If I reverse the predicate names, I need to explain this more thoroughly in comments and it looks complicated.

<details>
<summary>diff</summary>

```diff
diff --git a/src/box/lua/config/utils/schema.lua b/src/box/lua/config/utils/schema.lua
index 16f6c04dab..7bf588a437 100644
--- a/src/box/lua/config/utils/schema.lua
+++ b/src/box/lua/config/utils/schema.lua
@@ -139,9 +139,13 @@ end
 --     -- type otherwise.
 --     fromenv = <function>,
 --
---     -- If any number value definitely can't pass validation
---     -- against the given scalar, this property is true.
---     never_accept_number = <boolean>,
+--     -- If the given scalar type accepts some numeric values,
+--     -- this property is true.
+--     --
+--     -- It is not necessary that any number value is accepted.
+--     -- For example, if only integers are allowed, this property
+--     -- is true.
+--     accepts_number = <boolean>,
 -- }
 
 -- Verify whether the given value (data) has expected type and
@@ -182,7 +186,7 @@ scalars.string = {
     fromenv = function(_env_var_name, raw_value)
         return raw_value
     end,
-    never_accept_number = true,
+    accepts_number = false,
 }
 
 scalars.number = {
@@ -202,7 +206,7 @@ scalars.number = {
         end
         return res
     end,
-    never_accept_number = false,
+    accepts_number = true,
 }
 
 -- TODO: This hack is needed until a union schema node will be
@@ -215,7 +219,7 @@ scalars['string, number'] = {
     fromenv = function(_env_var_name, raw_value)
         return tonumber(raw_value) or raw_value
     end,
-    never_accept_number = false,
+    accepts_number = true,
 }
 scalars['number, string'] = {
     type = 'number, string',
@@ -225,7 +229,7 @@ scalars['number, string'] = {
     fromenv = function(_env_var_name, raw_value)
         return tonumber(raw_value) or raw_value
     end,
-    never_accept_number = false,
+    accepts_number = true,
 }
 
 scalars.integer = {
@@ -254,7 +258,7 @@ scalars.integer = {
         end
         return res
     end,
-    never_accept_number = false,
+    accepts_number = true,
 }
 
 scalars.boolean = {
@@ -276,7 +280,7 @@ scalars.boolean = {
         error(('Unable to decode a boolean value from environment ' ..
             'variable %q, got %q'):format(env_var_name, raw_value), 0)
     end,
-    never_accept_number = true,
+    accepts_number = false,
 }
 
 scalars.any = {
@@ -294,7 +298,7 @@ scalars.any = {
         end
         return res
     end,
-    never_accept_number = false,
+    accepts_number = true,
 }
 
 local function is_scalar(schema)
@@ -500,14 +504,14 @@ local function table_is_array(data)
     return true
 end
 
--- Whether the given record or map definitely can't accept a
--- non-empty array.
+-- Whether it is not guaranteed that the given record or map can't
+-- accept a non-empty array.
 --
--- Returns true if any non-empty array definitely can't pass the
+-- Returns false if any non-empty array definitely can't pass the
 -- schema's validation.
 --
--- Returns false if this property can't be guaranteed.
-local function never_accept_nonempty_array(schema)
+-- Returns true if this property can't be guaranteed.
+local function accepts_nonempty_array(schema)
     assert(schema.type == 'record' or schema.type == 'map')
 
     -- An array with a fixed max length may be represented by a
@@ -526,7 +530,7 @@ local function never_accept_nonempty_array(schema)
     --  |
     --  | s:validate({'a', 'b', 'c'}) -- OK
     if schema.type == 'record' then
-        return schema.fields[1] == nil
+        return schema.fields[1] ~= nil
     end
 
     -- If map's keys accept numeric values, then it is possible
@@ -542,11 +546,11 @@ local function never_accept_nonempty_array(schema)
     --  | s:validate({'a', 'b', 'c'}) -- OK
     assert(schema.type == 'map')
     if not is_scalar(schema.key) then
-        return true
+        return false
     end
     local scalar_def = scalars[schema.key.type]
     assert(scalar_def ~= nil)
-    return scalar_def.never_accept_number
+    return scalar_def.accepts_number
 end
 
 -- If a record or a map is known as discarding any non-empty
@@ -562,7 +566,7 @@ local function validate_for_unexpected_array(schema, data, ctx)
 
     -- If a non-empty array is a possibly valid value for the
     -- given schema, skip the check.
-    if not never_accept_nonempty_array(schema) then
+    if accepts_nonempty_array(schema) then
         return
     end
 
```
</details>

The 'never X' condition is what we actually need to know here, so I left it implemented this way. It seems the best we can do here. What do you think?

Reply:
I'm going to resolve the discussion, because Maxim assigns the pull request to me and I guess that he wants me to choose myself according to my perfonal feelings. OK.